src/Shared/TypeLoader.cs

```diff
@@ -47,7 +48,23 @@ internal class TypeLoader
 
         private static MetadataLoadContext _context;
 
-        private static string[] runtimeAssemblies = Directory.GetFiles(RuntimeEnvironment.GetRuntimeDirectory(), "*.dll");
+        private static string[] runtimeAssemblies = findRuntimeAssembliesWithMicrosoftBuildFramework();
+        private static string microsoftBuildFrameworkPath;
+
+        // We need to append Microsoft.Build.Framework from next to the executing assembly first to make sure it's loaded before the runtime variant.
+        private static string[] findRuntimeAssembliesWithMicrosoftBuildFramework()
+        {
+            string[] runtimeAssemblies = Directory.GetFiles(RuntimeEnvironment.GetRuntimeDirectory(), "*.dll");
+            string[] allAssemblies = new string[runtimeAssemblies.Length + 1];
+            microsoftBuildFrameworkPath = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), "Microsoft.Build.Framework.dll");
+            allAssemblies[0] = microsoftBuildFrameworkPath;
+            for (int i = 0; i < runtimeAssemblies.Length; i++)
+            {
+                allAssemblies[i + 1] = runtimeAssemblies[i];
+            }
+
+            return allAssemblies;
```

Comment:
This can be simplified:

```suggestion
            ((IList)runtimeAssemblies).Insert(0, microsoftBuildFrameworkPath);

            return runtimeAssemblies;
```